src/cargo/sources/config.rs

```diff
@@ -150,6 +156,32 @@ a lock file compatible with `{orig}` cannot be generated in this situation
             path.push(s);
             srcs.push(SourceId::for_directory(&path)?);
         }
+        if let Some(val) = table.get("git") {
+            let url = url(val, &format!("source.{}.git", name))?;
+            let try = |s: &str| {
+                let val = match table.get(s) {
+                    Some(s) => s,
+                    None => return Ok(None),
+                };
+                let key = format!("source.{}.{}", name, s);
+                val.string(&key).map(Some)
+            };
+            let reference = match try("branch")? {
+                Some(b) => GitReference::Branch(b.0.to_string()),
+                None => {
+                    match try("tag")? {
+                        Some(b) => GitReference::Tag(b.0.to_string()),
+                        None => {
+                            match try("rev")? {
+                                Some(b) => GitReference::Rev(b.0.to_string()),
+                                None => GitReference::Branch("master".to_string()),
```

Comment:
https://github.com/rust-lang/cargo/issues/3517

It's not fixed yet (I was not sure about it), so it's better to be consistent with our behavior elsewhere. 

Looks like the proper fix is to add a new `HEAD` variant to the `GitRefernce` enum, and it's to much work for this PR.